[Postgres] feature: Add Window Function Support For Query Generation

yaojiejia opened the pull request:
Implement window function support for PostgreSQL testing, including:
- Window function expressions with partition by, order by, row_number, rank, and frame clauses
- Named window definitions via window clause
- Integration with existing expression generation framework

The implementation allows SQLancer to generate and test queries containing window functions, expanding test coverage for PostgreSQL's analytical capabilities. Window functions are now included in the expression generation options and can be mutated during testing like other query components.

Key changes:
- Add PostgresWindowFunction and supporting classes
- Enhance PostgresSelect to handle window clauses
- Update expression generator for window function creation
- Modify visitor for proper SQL generation

In issue #76, it was mentioned that window functions were not planned for implementation due to lack of a test oracle. However, as a new contributor to open source, I wanted to take on this challenge and implement basic window function support for PostgreSQL testing. Please let me know if this function is still interesting or if I'm on the right track as I am very interested in tackling this challenge more. Thank you!

robins commented:
Thanks for working on this @yaojiejia !

I'll let others respond about the code, but +1 to adding support for this in SQLancer.

mrigger commented on src/sqlancer/postgres/gen/PostgresExpressionGenerator.java:
```diff
@@ -422,6 +422,53 @@ private PostgresExpression generateTextExpression(int depth) {
+        arguments.add(generateExpression(depth + 1));
+    }
+
+    // Generate PARTITION BY expressions
```
Instead of the comments, could we perhaps extract the relevant parts into methods with a descriptive name (e.g., here, we could have something like `generatePartitionByExpressions`?

mrigger commented on src/sqlancer/postgres/gen/PostgresExpressionGenerator.java:
```diff
@@ -667,7 +714,21 @@ public PostgresExpression generateBooleanExpression() {
+        PostgresSelect select = new PostgresSelect();
+        
+        // Add window functions to fetch columns if appropriate
+        if (Randomly.getBoolean()) {
```
Could we set a lower probability for this? I remember that we have a method in the `Randomly` class for that.

mrigger commented on src/sqlancer/postgres/gen/PostgresWindowFunctionGenerator.java:
```diff
@@ -0,0 +1,96 @@
+        String functionName = Randomly.fromList(WINDOW_FUNCTIONS);
+        List<PostgresExpression> arguments = new ArrayList<>();
+        
+        // Generate function arguments based on function name
```
Also here and below, I think it would improve readability to extract common blocks into methods instead of using comments.

mrigger reviewed (commented):
Thanks a lot for the PR! This is great. I have some minor improvement suggestions, and we can merge the PR after that.

yaojiejia commented:
Hi @mrigger , just refactored the code with the proper descriptive name and changed the probability of the window function generator, let me know if there are any other things I should look at! Thank you for your time!

mrigger reviewed (approved):
Thanks a lot again for the PR!

mrigger commented:
It seems the tests fail after applying the last two commits. If you undo them, I think I can directly merge the PR.

yaojiejia commented:
sorry about that, will undo them now

yaojiejia commented:
just revert it back, let me know if you need anything :)

mrigger commented:
Hmm, it seems there is some compilation issue now.

yaojiejia commented:
Hey @mrigger just fixed the compilation error and it is successfully running on my local

mrigger commented:
Thanks! Currently, the style checks fail. You can fix this by executing `mvn formatter:format`. You can also double-check by using `mvn verify -DskipTests=true`.

yaojiejia commented:
Just fixed the style checks issue, let me know if there are other things I should be looking at @mrigger 
thanks!

mrigger commented on src/sqlancer/postgres/PostgresToStringVisitor.java:
```diff
@@ -362,4 +372,61 @@ public void visit(PostgresLikeOperation op) {
+    public void visit(PostgresWindowFunction windowFunction) {
+        sb.append(windowFunction.getFunctionName());
+        sb.append("(");
+        // Fix: Use visitList instead of visit for a list of expressions
```
We can omit the comment here.

mrigger commented on src/sqlancer/postgres/PostgresToStringVisitor.java:
```diff
@@ -362,4 +372,61 @@ public void visit(PostgresLikeOperation op) {
+                sb.append(" ");
+            }
+            sb.append("ORDER BY ");
+            // Fix: Create a method to handle order by terms specifically
```
We can omit the comment here.

mrigger commented on src/sqlancer/postgres/PostgresToStringVisitor.java:
```diff
@@ -362,4 +372,61 @@ public void visit(PostgresLikeOperation op) {
+    }
+
+    // this method handles lists of expressions
+    private void visitList(List<PostgresExpression> expressions) {
```
The `ToStringVisitor` superclass already has a method to visit a list of expressions, which we can use.

mrigger commented on src/sqlancer/postgres/PostgresToStringVisitor.java:
```diff
@@ -362,4 +372,61 @@ public void visit(PostgresLikeOperation op) {
+    }
+
+    // this method handles lists of order by terms
+    private void visitOrderByList(List<PostgresOrderByTerm> orderBy) {
```
`PostgresOrderByTerm` is a `PostgresExpression`, so I think we don't need a separate method here.

mrigger commented on src/sqlancer/postgres/gen/PostgresExpressionGenerator.java:
```diff
@@ -52,7 +52,7 @@
 import sqlancer.postgres.ast.PostgresLikeOperation;
 import sqlancer.postgres.ast.PostgresOrderByTerm;
-import sqlancer.postgres.ast.PostgresOrderByTerm.PostgresOrder;
+// import sqlancer.postgres.ast.PostgresOrderByTerm.PostgresOrder;
```
We can remove this.

mrigger commented on src/sqlancer/postgres/gen/PostgresWindowFunctionGenerator.java:
```diff
@@ -0,0 +1,138 @@
+    private static final List<String> WINDOW_FUNCTIONS = Arrays.asList("row_number", "rank", "dense_rank",
+            "percent_rank", "cume_dist", "ntile", "lag", "lead", "first_value", "last_value", "nth_value");
+
+    // Private constructor to prevent instantiation
```
We can omit the comment.

mrigger commented on src/sqlancer/postgres/gen/PostgresWindowFunctionGenerator.java:
```diff
@@ -0,0 +1,138 @@
+        return partitionBy;
+    }
+
+    private static List<PostgresOrderByTerm> generateOrderByClause(List<PostgresExpression> availableExpr) {
```
We already have a method in the `PostgresExpressionGenerator`, so I think we should use that one.

mrigger commented on src/sqlancer/postgres/PostgresToStringVisitor.java:
```diff
@@ -185,11 +189,17 @@ public void visit(PostgresSelect s) {
         }
     }
 
+    // @Override
```
We should omit commented-out code.

mrigger reviewed (commented):
Thanks! While having a look again at the code, I identified some further improvement opportunities that I overlooked. Could you please have another look at them?

yaojiejia commented:
Thanks for the feedback and I just fixed them!

mrigger commented:
Thanks! Just wanted to merge, but noticed that the current version causes some syntax errors such as the following:

> --java.lang.AssertionError: CREATE INDEX i0 ON t2((num_nonnulls(- (length((t2.c0 COLLATE "C"))), ((+ (2004518786))^(- (-835456169))), - (last_value() OVER ( RANGE BETWEEN -1575743297 AND -1092944323)))) DESC, c0 ASC);
> --	at sqlancer.common.query.SQLQueryAdapter.checkException(SQLQueryAdapter.java:124)
> --	at sqlancer.common.query.SQLQueryAdapter.execute(SQLQueryAdapter.java:106)
> --	at sqlancer.Main$QueryManager.execute(Main.java:355)
> --	at sqlancer.GlobalState.executeStatement(GlobalState.java:108)
> --	at sqlancer.StatementExecutor.executeStatements(StatementExecutor.java:70)
> --	at sqlancer.postgres.PostgresProvider.prepareTables(PostgresProvider.java:328)
> --	at sqlancer.postgres.PostgresProvider.generateDatabase(PostgresProvider.java:206)
> --	at sqlancer.postgres.PostgresProvider.generateDatabase(PostgresProvider.java:1)
> --	at sqlancer.ProviderAdapter.generateAndTestDatabase(ProviderAdapter.java:52)
> --	at sqlancer.Main$DBMSExecutor.run(Main.java:454)
> --	at sqlancer.Main$2.run(Main.java:668)
> --	at sqlancer.Main$2.runThread(Main.java:650)
> --	at sqlancer.Main$2.run(Main.java:641)
> --	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> --	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> --	at java.base/java.lang.Thread.run(Thread.java:829)

yaojiejia commented:
Ahh i see the issue, the PostgresWindowFunctionGenerator class is generating window functions, but there's no validation to ensure they're only used in contexts where postgresql allows them. Window functions are primarily meant for select statements, but sqlancer is trying to use them in many other contexts. I think this might take a little bit of time to fix but I will continuously work on this issue, thanks for being patient!
